src/distributions/normal.jl

```diff
@@ -538,37 +538,45 @@ end
 
 # Thes functions extends the `CVI` approximation method in case if input is from the `NormalDistributionsFamily`
 
-get_df_m(::Type{<:UnivariateNormalNaturalParameters}, ::Type{<:UnivariateGaussianDistributionsFamily}, logp_nc::Function) = (z) -> ForwardDiff.derivative(logp_nc, z)
-
-get_df_m(::Type{<:MultivariateNormalNaturalParameters}, ::Type{<:MultivariateGaussianDistributionsFamily}, logp_nc::Function) = (z) -> ForwardDiff.gradient(logp_nc, z)
+function compute_second_derivative(grad::G, logp::F, z_s::Real) where {G, F}
+    first_derivative = (x) -> compute_derivative(grad, logp, x)
+    return compute_derivative(grad, first_derivative, z_s)
+end
 
-get_df_v(::Type{<:UnivariateNormalNaturalParameters}, ::Type{<:UnivariateGaussianDistributionsFamily}, df_m::Function) = (z) -> ForwardDiff.derivative(df_m, z)
+function compute_df_mv(approximation::CVI, logp::F, z_s::Real) where {F}
+    df_m = compute_derivative(get_grad(approximation), logp, z_s)
+    df_v = compute_second_derivative(get_grad(approximation), logp, z_s)
+    return df_m, df_v / 2
+end
 
-get_df_v(::Type{<:MultivariateNormalNaturalParameters}, ::Type{<:MultivariateGaussianDistributionsFamily}, df_m::Function) = (z) -> ForwardDiff.jacobian(df_m, z)
+function compute_df_mv(approximation::CVI, logp_nc::F, z_s::Vector) where {F}
+    df_m = compute_gradient(get_grad(approximation), logp_nc, z_s)
+    df_v = compute_hessian(get_grad(approximation), logp_nc, z_s)
+    return df_m, df_v ./ 2
+end
 
-function render_cvi(approximation::CVIApproximation, logp_nc::F, initial::GaussianDistributionsFamily) where {F}
-    η = naturalparams(initial)
-    λ = naturalparams(initial)
+function prod(approximation::CVI, logp::F, dist::GaussianDistributionsFamily) where {F <: Function}
+    η = naturalparams(dist)
+    λ = naturalparams(dist)
     T = typeof(η)
 
     rng = something(approximation.rng, Random.GLOBAL_RNG)
     opt = approximation.opt
     its = approximation.num_iterations
 
-    df_m = (z) -> get_df_m(typeof(λ), typeof(initial), logp_nc)(z)
-    df_v = (z) -> get_df_v(typeof(λ), typeof(initial), df_m)(z) / 2
-
     hasupdated = false
 
     for _ in 1:its
         q = convert(Distribution, λ)
         z_s = rand(rng, q)
-        df_μ1 = df_m(z_s) - 2 * df_v(z_s) * mean(q)
-        df_μ2 = df_v(z_s)
+
+        df_m, df_v = compute_df_mv(approximation, logp, z_s)
+        df_μ1 = df_m - 2 * df_v * mean(q)
+        df_μ2 = df_v
         ∇f = as_naturalparams(T, df_μ1, df_μ2)
         ∇ = λ - η - ∇f
         λ_new = as_naturalparams(T, cvi_update!(opt, λ, ∇))
-        if isproper(λ_new)
+        if isproper(λ_new) && enforce_proper_message(approximation.enforce_proper_messages, λ_new, η)
             λ = λ_new
             hasupdated = true
         end
```

Comment:
Let's add some more comments here, similarly as in `cvi.jl`

Reply:
[Done](https://github.com/biaslab/ReactiveMP.jl/pull/239/commits/2ed0b6e3cc56db2ea21cbe97e4945529e3b3613e)